### ISSUE TITLE
NULL should be used instead of 0 for pointer values

### ISSUE THREAD

**Issue by thebombzen** [libjxl/libjxl#610]
**Describe the bug**
The C API has some places where `0` is returned instead of `NULL` for pointer values. One such example is <https://libjxl.readthedocs.io/en/latest/api_common.html#_CPPv417jpegxl_alloc_func> [(internet archive)](https://web.archive.org/web/20210917203032/https://libjxl.readthedocs.io/en/latest/api_common.html). According to the [C99 spec](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) [(internet archive)](https://web.archive.org/web/20210911061459/http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf), NULL (defined in stddef.h) is required to be a null pointer, and null pointers are required to cast from numerical zero, but null pointers are not required to actually equal numerical zero. That is, the spec requires that `NULL == (void *)0` but the spec does not actually require that `NULL == 0`. (The C++ spec requires this, but the API is C.)

**Expected behavior**
NULL should be used in all cases where a null pointer value is required, not numerical zero, for full strict portability. I'm not aware of any implementation where NULL is not numerically zero, but the spec allows for it.

**Comment by deymo** [comment]
The functions in the C API are all implemented in C++, so I don't think this is an undefined behavior problem, but the documentation should be updated.

**Comment by thebombzen** [comment]
If they do actually return `NULL` and not numerical zero then the documentation should be updated to reflect this, yes.

**Comment by TheHardew** [comment]
The return type of `jpegxl_alloc_func` is `void*`.
I believe in C `(void*)0` always returns `NULL` , even if `NULL` is not actually 0, e.g. `(void*) 0 == (void*) -1` would be true if `NULL` was -1.

**Comment by thebombzen** [comment]
@TheHardew I did specifically mention all of this in the opened issue, btw, but let me repeat it here. The C spec does not guarantee that `NULL == 0` but it does guarantee that `NULL == (void *)0`.

**Comment by Hello71** [comment]
You are correct that C does not require null pointers to be represented as all-bits-zero. However, it is not correct that you must use the built-in macro NULL in order to obtain a null pointer constant. In fact, NULL is not even guaranteed to be of pointer type in ISO C. From https://www.iso-9899.info/n1570.html: 7.19(3): "NULL [...] expands to an implementation-defined null pointer constant", 6.3.2.3(3): "An integer constant expression with the value 0, or such an expression cast to type void *, is called a null pointer constant". Many implementations choose to define NULL as 0 or some other uncasted integer constant expression, such as musl: https://ewontfix.com/11/.

Even so, `NULL == 0` is guaranteed by the ISO C standard to evaluate to 1. If NULL is of integer type, then 6.5.9(4): "If both of the operands have arithmetic type, the usual arithmetic conversions are performed." and obviously `0 == 0` is true. If NULL is of pointer type, then 6.5.9(5): "If one operand is a pointer and the other is a null pointer constant, the null pointer constant is converted to the type of the pointer.", 6.5.9(6): "Two pointers compare equal if and only if both are null pointers, [...]". And for completeness, 6.5.9(3): "Each of the operators yields 1 if the specified relation is true and 0 if it is false." (in case someone tries to claim that == may return a value other than 1 or 0).

It is true that `int p = 0; (void*) p;` is implementation-defined, but hopefully nobody would write such code anyways, since setting p = any non-zero value would have unexpected behavior, and if you only ever set `p = 0` then you may as well inline the definition (or use a macro).

In fact, since the C standard has such loose requirements on the definition of NULL, it is not even sufficient in some cases. The most common case, as described at https://ewontfix.com/11/, is when invoking variadic functions such as `execl` [0], which, in POSIX, must be terminated by `(char *)0`, not by `NULL`. See https://man7.org/linux/man-pages/man3/exec.3.html "this pointer must be cast `(char *) NULL`", https://pubs.opengroup.org/onlinepubs/9699919799/functions/exec.html which has `(char *)0` everywhere instead of `NULL`. Some programmers, including Rich Felker, the author of the excellent musl libc, prefer to use a simple 0 in cases where it is guaranteed to be converted to a null pointer constant, including when calling a non-variadic function accepting a pointer type, or returning from a function defined as returning a pointer type.

In conclusion, while it may improve the clarity of the documentation to replace 0 with NULL, it is by no means required for correctness.

[0] which is not part of ISO C but can be used as an example of a variadic function

**Comment by Hello71** [comment]
Regarding nullptr, it is definitely better than NULL for use in C++11 code. As far as I can tell, though, libjxl only uses NULL in C or when interfacing with C, and consistently uses nullptr elsewhere. The C interfacing parts could be converted to nullptr but with negligible benefit. Regarding "it can be wrongly redefined", that is true of anything; you may legally `#define nullptr abort()` in C++. Hell, you can even do nonsense things like `#define if while`, but nobody argues that we should therefore avoid if.

Also, C++ technically does require `NULL == 0`, but only in the same way as C, not because it's necessarily an integer constant. In C++, `NULL` can be defined as `nullptr`, which is not an integer.
